semtypes/src/main/java/io/ballerina/types/typeops/ListCommonOps.java

```diff
@@ -118,82 +119,154 @@ static boolean listFormulaIsEmpty(Context cx, Conjunction pos, Conjunction neg)
+        return !listInhabited(cx, indices.toArray(new Integer[0]), sampleTypes.memberTypes.toArray(new SemType[0]),
+                sampleTypes.nRequired, neg);
     }
 
-    static FixedLengthArraySemtypePair listIntersectWith(FixedLengthArray members,
-                                                         SemType rest, ListAtomicType newType) {
-        int newTypeLen = newType.members.fixedLength;
-        int intersectedLen = Integer.max(members.fixedLength, newTypeLen);
-        // We can specifically handle the case where length of `initial` and `fixedLength` are the same
-        if (members.fixedLength < intersectedLen) {
-            if (Core.isNever(rest)) {
-                return null;
+    public static SemtypesIntPair listSampleTypes(Context cx, FixedLengthArray members, SemType rest,
+                                                   List<Integer> indices) {
+        List<SemType> memberTypes = new ArrayList<>();
+        int nRequired = 0;
+        for (int i = 0; i < indices.size(); i++) {
+            int index = indices.get(i);
+            SemType t = listMemberAt(members, rest, index);
+            if (Core.isEmpty(cx, t)) {
+                break;
+            }
+            memberTypes.add(t);
+            if (index < members.fixedLength) {
+                nRequired = i + 1;
             }
-            fixedArrayFill(members, intersectedLen, rest);
         }
-        int maxInitialLen = Integer.max(members.initial.size(), newType.members.initial.size());
-        for (int i = 0; i < maxInitialLen; i++) {
-            fixedArraySet(members,
-                          i,
-                          intersect(listMemberAt(members, rest, i), listMemberAt(newType.members, newType.rest, i)));
+        // Note that indices may be longer
+        return SemtypesIntPair.from(memberTypes, nRequired);
+    }
+
+    // Return a list of sample indices for use as second argument of `listInhabited`.
+    // The positive list type P is represented by `members` and `rest`.
+    // The negative list types N are represented by `neg`
+    // The `indices` list (first member of return value) is constructed in two stages.
+    // First, the set of all non-negative integers is partitioned so that two integers are
+    // in different partitions if they get different types as an index in P or N.
+    // Second, we choose a number of samples from each partition. It doesn't matter
+    // which sample we choose, but (this is the key point) we need at least as many samples
+    // as there are negatives in N, so that for each negative we can freely choose a type for the sample
+    // to avoid being matched by that negative.
+    public static List<Integer> listSamples(Context cx, FixedLengthArray members, SemType rest, Conjunction neg) {
+        int maxInitialLength = members.initial.size();
+        List<Integer> fixedLengths = new ArrayList<>(members.fixedLength);
```

Comment:
`int[] fixedLengths = [members.fixedLength];` -> `fixedLengths.add(members.fixedLength)`

Reply:
This could be the bug that causes unit test failures